### PR TITLE
Rewrite the Cargo.lock tidy check.

### LINKED ISSUE #8691: tidy.py check for cargo versions doesn't report multiple byteorder packages
Noticed in review for #8446.

### PR DESCRIPTION
**Pull request by Ms2ger**
Fixes #8691.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8824)

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 188 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8824#-K4hQjIxdxKs1H_gWrwP:-K4hQjIxdxKs1H_gWrwQ:525476749) ([raw file](https://github.com/servo/servo/blob/172978498604e36649780fa3146542f9161fa33b/python/tidy.py#L188)):*
The first argument of yield should be the line number, no ?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8824)*

**Comment by Ms2ger** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[python/tidy.py, line 188 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8824#-K4hQjIxdxKs1H_gWrwP:-K4hUwMnGb1hQL0wBfLE:-917290107) ([raw file](https://github.com/servo/servo/blob/172978498604e36649780fa3146542f9161fa33b/python/tidy.py#L188)):*
Yes, but I don't have the line number at this point

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8824)*

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 188 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8824#-K4hQjIxdxKs1H_gWrwP:-K4hVGCcU_PWTKPePQ6s:-2118747493) ([raw file](https://github.com/servo/servo/blob/172978498604e36649780fa3146542f9161fa33b/python/tidy.py#L188)):*
It's a problem since there are a lot of place where package names are present.  :-/

Isn't there a way to keep a track of the lines ?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8824)*

**Comment by jdm** [comment]
That might be the reason we didn't use the TOML parsing library in the beginning.

**Comment by Ms2ger** [comment]
```
./components/servo/Cargo.lock:1: duplicate versions for package "libc"
	found dependency on version 0.1.12
	but highest version is 0.2.2
	try upgrading with ./mach cargo-update -p libc:0.1.12
	The following packages depend on version 0.1.12:
		block
		clock_ticks
		dlib
		energymon
		energymon-default-sys
		energymon-sys
		gl_common
		hbs-common-sys
		hbs-pow
		hbs-pow-sys
		libz-sys
		malloc_buf
		memchr
		mmap
		num_cpus
		objc-foundation
		wayland-client
		x11-dl
./components/servo/Cargo.lock:1: duplicate versions for package "byteorder"
	found dependency on version 0.3.13
	but highest version is 0.4.2
	try upgrading with ./mach cargo-update -p byteorder:0.3.13
	The following packages depend on version 0.3.13:
		bincode
./ports/gonk/Cargo.lock:1: duplicate versions for package "libc"
	found dependency on version 0.1.12
	but highest version is 0.2.2
	try upgrading with ./mach cargo-update -p libc:0.1.12
	The following packages depend on version 0.1.12:
		block
		clock_ticks
		dlib
		gl_common
		hbs-common-sys
		hbs-pow
		hbs-pow-sys
		libz-sys
		malloc_buf
		memchr
		mmap
		num_cpus
		objc-foundation
		wayland-client
		x11-dl
./ports/gonk/Cargo.lock:1: duplicate versions for package "byteorder"
	found dependency on version 0.3.13
	but highest version is 0.4.2
	try upgrading with ./mach cargo-update -p byteorder:0.3.13
	The following packages depend on version 0.3.13:
		bincode
./ports/cef/Cargo.lock:1: duplicate versions for package "libc"
	found dependency on version 0.1.12
	but highest version is 0.2.2
	try upgrading with ./mach cargo-update -p libc:0.1.12
	The following packages depend on version 0.1.12:
		block
		clock_ticks
		dlib
		gl_common
		hbs-common-sys
		hbs-pow
		hbs-pow-sys
		libz-sys
		malloc_buf
		memchr
		mmap
		num_cpus
		objc-foundation
		wayland-client
		x11-dl
./ports/cef/Cargo.lock:1: duplicate versions for package "cocoa"
	found dependency on version 0.1.5
	but highest version is 0.2.2
	try upgrading with ./mach cargo-update -p cocoa:0.1.5
	The following packages depend on version 0.1.5:
		embedding
./ports/cef/Cargo.lock:1: duplicate versions for package "byteorder"
	found dependency on version 0.3.13
	but highest version is 0.4.2
	try upgrading with ./mach cargo-update -p byteorder:0.3.13
	The following packages depend on version 0.3.13:
		bincode
```

**Comment by Wafflespeanut** [comment]
Though I agree that the line numbers are very helpful while hand-editing the lock files, I'm very happy with the latest changes, which provide enough information about which packages we should be considering for an update. @Ms2ger It's beautiful! :)

File: python/tidy.py [hunk 1]
```diff
@@ -159,45 +159,52 @@ def stdout_redirect(where):
...
-                    yield (line, message + error + suggest)
-                idx += 1
-        idx += 1
+    exceptions = ["libc", "byteorder", "cocoa"]
+
+    import toml
+    content = toml.loads(contents)
+
+    packages = {}
+    for package in content.get("package", []):
```
Comment (Wafflespeanut) [inline python/tidy.py]: Nit: Shall we have the import block at the top along with the others? Also, separate the declarations from the iterations? (for the sake of beauty! :P)

Comment (Wafflespeanut) [inline python/tidy.py]: never mind...

**Comment by Wafflespeanut** [comment]
The changes look great, r=me

**Pull request merged**